src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -598,6 +604,34 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
             }
         }
 
+        private void VerifyWrite()
+        {
+            TryInitializeResponseContentLength();
+
+            if (_responseContentLength.HasValue && _responseBytesWritten > _responseContentLength.Value)
+            {
+                ThrowContentLengthMismatchException();
+            }
+        }
+
+        protected void TryInitializeResponseContentLength()
+        {
+            if (!_responseContentLength.HasValue)
+            {
+                var responseHeaders = FrameResponseHeaders;
+                if (responseHeaders != null && responseHeaders.HasContentLength)
+                {
+                    _responseContentLength = int.Parse(responseHeaders.HeaderContentLength.ToString());
```

Comment:
I'll leave it as `Parse` since that will throw if the app does something like `                httpContext.Response.Headers["Content-Length"] = "foo";`